### PR TITLE
cmake: Add external signer support

### PR DESCRIPTION
**Pull request by hebasto**
A new configuration option is `WITH_EXTERNAL_SIGNER`.

### CODE REVIEW

File: vcpkg.json [hunk 1]
```diff
@@ -5,7 +5,6 @@
     "berkeleydb",
     "boost-date-time",
     "boost-multi-index",
-    "boost-process",
```
[the remark this answers is not in the document]
Comment (hebasto) [reply vcpkg.json:8]: It mirrors https://github.com/bitcoin/bitcoin/pull/28967.

**Comment by hebasto** [comment]
Rebased.

**Comment by fanquake** [comment]
> This branch works without the following workarounds from the master branch: 

How is the first one fixed? I'm guessing it's because these checks are no-longer being run with same warning flags? I assume the second one is because -pthread is now just always being passed or something?

**Comment by hebasto** [comment]
> > This branch works without the following workarounds from the master branch:
> 
> How is the first one fixed? I'm guessing it's because these checks are no-longer being run with same warning flags?

Thanks! Fixed.

**Comment by hebasto** [comment]
> I assume the second one is because -pthread is now just always being passed or something?

Not so simple :)

I tried to reproduce the initial issue, which was fixed in https://github.com/bitcoin/bitcoin/pull/24397, using for testing the [077cfffebd1b12ec4e2efb949fd9a79fb5be3748]([url](https://github.com/bitcoin/bitcoin/commit/077cfffebd1b12ec4e2efb949fd9a79fb5be3748)) commit and commands as follows:
```sh
make -C depends -j $(nproc) NO_QT=1 NO_WALLET=1 NO_ZMQ=1 NO_UPNP=1 NO_NATPMP=1 NO_USDT=1 CC=gcc-10 CXX=g++-10
./autogen.sh && ./configure CONFIG_SITE=$PWD/depends/x86_64-pc-linux-gnu/share/config.site --enable-external-signer
```

1. Ubuntu 20.04, libc 2.31 -- FAIL

config.log:
```
configure:26954: checking whether Boost.Process can be used
configure:26967: g++-10 -std=c++17 -o conftest -pipe -O2  -fno-extended-identifiers -I/bitcoin/depends/x86_64-pc-linux-gnu/include/  -DHAVE_BUILD_INFO -DPROVIDE_FUZZ_MAIN_FUNCTION -L/bitcoin/depends/x86_64-pc-linux-gnu/lib  conftest.cpp  >&5
/usr/bin/ld: /tmp/ccTX7pPT.o: in function `boost::asio::detail::posix_global_impl<boost::asio::system_context>::~posix_global_impl()':
conftest.cpp:(.text._ZN5boost4asio6detail17posix_global_implINS0_14system_contextEED2Ev[_ZN5boost4asio6detail17posix_global_implINS0_14system_contextEED5Ev]+0xa3): undefined reference to `pthread_join'
/usr/bin/ld: conftest.cpp:(.text._ZN5boost4asio6detail17posix_global_implINS0_14system_contextEED2Ev[_ZN5boost4asio6detail17posix_global_implINS0_14system_contextEED5Ev]+0xc4): undefined reference to `pthread_detach'
collect2: error: ld returned 1 exit status
```

2. Ubuntu 22.04, libc 2.35 -- SUCCESS

FWIW, reading libc release notes did not give a clue.

**Comment by fanquake** [comment]
I would guess that's because recent versions of libc now include pthread, so pthread linking is no-longer required. I don't understand how this branch can drop the workaround in other cases, unless pthread is being linked some other way.

**Comment by fanquake** [comment]
> FWIW, reading libc release notes did not give a clue.

It's in the 2.34 release notes: https://sourceware.org/pipermail/libc-alpha/2021-August/129718.html

> In order to support smoother in-place-upgrades and to simplify
  the implementation of the runtime all functionality formerly
  implemented in the libraries libpthread, libdl, libutil, libanl has
  been integrated into libc.  New applications do not need to link with
  -lpthread, -ldl, -lutil, -lanl anymore.

File: cmake/module/AddBoostIfNeeded.cmake [hunk 2]
```diff
@@ -48,5 +49,53 @@ function(add_boost_if_needed)
...
   endif()
 
-  mark_as_advanced(Boost_INCLUDE_DIR)
+  if(WITH_EXTERNAL_SIGNER)
+    include(CheckCXXSourceCompiles)
+    set(CMAKE_REQUIRED_INCLUDES ${Boost_INCLUDE_DIR})
+    set(CMAKE_REQUIRED_FLAGS "${working_compiler_werror_flag}")
+    # Boost 1.78 requires the following workaround.
+    # See: https://github.com/boostorg/process/issues/235
+    string(APPEND CMAKE_REQUIRED_FLAGS " -Wno-error=narrowing")
```
Comment (hebasto) [inline cmake/module/AddBoostIfNeeded.cmake:58]: By default, CMake [treats](https://cmake.org/cmake/help/latest/prop_tgt/NO_SYSTEM_FROM_IMPORTED.html)
> include directories from the interfaces of consumed [Imported Targets](https://cmake.org/cmake/help/latest/manual/cmake-buildsystem.7.html#imported-targets) as system directories.

That means that we have `--enable-suppress-external-warnings` behaviour for granted.

If we won't bother to mirror the opposite behaviour, these 4 lines might be dropped.

**Comment by fanquake** [comment]
PR description needs updating?

**Comment by hebasto** [comment]
> PR description needs updating?

Sure. Updated.

**Comment by theuni** [comment]
Looks good to me other than the outstanding comments:
- Werror question
- @fanquake's request for a better comment re glibc.

File: cmake/module/AddBoostIfNeeded.cmake [hunk 3]
```diff
@@ -48,5 +49,53 @@ function(add_boost_if_needed)
     endif()
   endif()
 
-  mark_as_advanced(Boost_INCLUDE_DIR)
+  if(WITH_EXTERNAL_SIGNER)
+    include(CheckCXXSourceCompiles)
+    set(CMAKE_REQUIRED_INCLUDES ${Boost_INCLUDE_DIR})
+    set(CMAKE_REQUIRED_FLAGS "${working_compiler_werror_flag}")
```
[the remark this answers is not in the document]
Comment (hebasto) [reply cmake/module/AddBoostIfNeeded.cmake]: Yes, it does not correlate with the master branch (and I don't recall why I added it).

This line has been removed in the recent push to mirror the master branch behavior.

**Comment by hebasto** [comment]
Thanks for review!

> Looks good to me other than the outstanding comments:
> 
>    * Werror question
> 
>    * @fanquake's request for a better comment re glibc.

Updated per feedback.

**Comment by m3dwards** [comment]
ACK https://github.com/hebasto/bitcoin/pull/79/commits/617ef5b5a9532903630b0bcbaa74ed54c267709c

Copied @pablomartin4btc trick for checking if signer support is available.

With `cmake -S .. -DWITH_EXTERNAL_SIGNER=off`:

```shell
➜  build git:(240118-cmake-AY) ./src/bitcoin-cli -regtest -datadir=/tmp/bitcoinddata enumeratesigners
error code: -32601
error message:
Method not found
```

With `cmake -S .. -DWITH_EXTERNAL_SIGNER=on`:

```shell
➜  build git:(240118-cmake-AY) ./src/bitcoin-cli -regtest -datadir=/tmp/bitcoinddata enumeratesigners
error code: -1
error message:
Error: restart bitcoind with -signer=<cmd>
```

----

Just as an aside, when I set external signer to "on" lowercase the print out after configure is: `external signer ..................... ON` in uppercase but when I set it to `off` lowercase the print out after configure is: `external signer ..................... off` in lowercase. Is this a cmake quirk?

File: cmake/module/AddBoostIfNeeded.cmake [hunk 4]
```diff
@@ -54,5 +55,54 @@ function(add_boost_if_needed)
...
+        bp::child c(\"dummy\", bp::std_out > stdout_stream, bp::std_err > stdout_stream, bp::std_in < stdin_stream);
+        stdin_stream << std::string{\"test\"} << std::endl;
+        if (c.running()) c.terminate();
+        c.wait();
+        c.exit_code();
+      }
+      " HAVE_BOOST_PROCESS
+    )
+    if(HAVE_BOOST_PROCESS)
+      if(WIN32)
```
Comment (m3dwards) [inline cmake/module/AddBoostIfNeeded.cmake:87]: NIT: if it's WIN32 couldn't you just return early with external signer OFF and not bother checking if boost supports boost process? Would cleanup one nested if.

Comment (hebasto) [reply cmake/module/AddBoostIfNeeded.cmake:87]: You are right. However, I've chosen this clumsy approach to mimic the Autotools' one for the sake of easier reviewing,

Anyway, this code will gone after https://github.com/bitcoin/bitcoin/pull/28981.

**Comment by hebasto** [comment]
@m3dwards 
> Just as an aside, when I set external signer to "on" lowercase the print out after configure is: `external signer ..................... ON` in uppercase but when I set it to `off` lowercase the print out after configure is: `external signer ..................... off` in lowercase. Is this a cmake quirk?

That is a flaw in tri-state option implementation. Anyway, it will be gone before merging into the master branch; see https://github.com/hebasto/bitcoin/pull/83.

**Comment by pablomartin4btc** [comment]
I just found out that this feature is disabled for Windows build in `master`. For some more context see @Sjors [comment](https://github.com/bitcoin/bitcoin/pull/28967#issuecomment-1966952430).

**Comment by Sjors** [comment]
@pablomartin4btc indeed, though after #28981 bringing back Windows should be doable.

**Pull request merged**